### PR TITLE
TEP-0090: Matrix - API Change and Feature Flag

### PR DESCRIPTION
**Pull request by jerop**
# Changes

[TEP-0090][tep-0090] proposed executing a `PipelineTask` in parallel `TaskRuns` and `Runs` with substitutions from combinations of `Parameters` in a `Matrix`.

In this change, we add the `matrix` field to the `PipelineTask` specification, which takes a list of `Parameters`. We gate this field behind the `alpha` feature gate. We add validation and testing that the field is available in `alpha` only - we will add further validation in upcoming changes. We also add documentation the API change - we will incrementally add further details as we iterate on the implementation of [TEP-0090: Matrix][tep-0090].

[tep-0090]: https://github.com/tektoncd/community/blob/main/teps/0090-matrix.md

/kind feature

Issue: https://github.com/tektoncd/pipeline/issues/2050

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

```release-note
Added syntax support for `Matrix` as an alpha feature. It is still in a very early stage of development and is not yet fully functional.
```

### CODE REVIEW

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go [hunk 1]
```diff
@@ -546,3 +549,14 @@ func validateGraph(tasks []PipelineTask) *apis.FieldError {
 	}
 	return nil
 }
+
+func validateMatrix(ctx context.Context, tasks []PipelineTask) (errs *apis.FieldError) {
```
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/pipeline_validation.go:553]: Does this require additional validation like the functionality provided [here](https://github.com/tektoncd/pipeline/blob/99b8b196ea753af36befda8c0e0e1eaa9490ae68/pkg/apis/pipeline/v1beta1/pipeline_validation.go#L132)? e.g. validating param types and defaults?

Comment (jerop) [reply pkg/apis/pipeline/v1beta1/pipeline_validation.go:553]: Yes, will add those in the next step because it's a lot and want to implement this incrementally (targeting all validation to be included in the next release though)

As noted in the commit and PR message:
> We add validation and testing that the field is available in `alpha` only - we will add further validation in upcoming changes.

**Comment by jerop** [comment]
[`TestSidecarTaskSupport`](https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_pipeline/4600/pull-tekton-pipeline-integration-tests/1496194554930401283/#test-testsidecartasksupport) is a known flake reported in https://github.com/tektoncd/pipeline/issues/2656

/test pull-tekton-pipeline-integration-tests

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go (hunk 1 above)
Comment (lbernick) [reply pkg/apis/pipeline/v1beta1/pipeline_validation.go:553]: whoops my bad :)

Comment (jerop) [reply pkg/apis/pipeline/v1beta1/pipeline_validation.go:553]: no worries, thanks for the review 😀

**Comment by jerop** [comment]
@tektoncd/core-maintainers looking for reviews, hoping open one or two more PRs adding more validation before the next release 🙏🏾

File: docs/matrix.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<!--
```
Comment (pritidesai) [inline docs/matrix.md:1]: copyright license please 🙏

File: docs/matrix.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+## Overview
+
+`Matrix` is used to fan out `Tasks` in a `Pipeline`. This doc will explain the details of `matrix` support in
+Tekton. 
+
+> :seedling: **`Matrix` is an [alpha](install.md#alpha-features) feature.**
+> The `enable-api-fields` feature flag must be set to `"alpha"` to specify `Matrix` in a `PipelineTask`.
+>
+> :warning: This feature is still in very early stage of development and is not yet functional. Do not use it.
```
Comment (pritidesai) [inline docs/matrix.md:20]: suggest rewording to:

> :warning: This feature is in a preview mode. It is still in a very early stage of development and is not yet fully functional.

File: docs/pipelines.md [hunk 4]
```diff
@@ -307,6 +312,30 @@ spec:
...
+### Specifying `Matrix` in `PipelineTasks`
+
+> :seedling: **`Matrix` is an [alpha](install.md#alpha-features) feature.**
+> The `enable-api-fields` feature flag must be set to `"alpha"` to specify `Matrix` in a `PipelineTask`.
+>
+> :warning: This feature is still in very early stage of development and is not yet functional. Do not use it.
+
+You can also provide [`Parameters`](tasks.md#specifying-parameters) through the `matrix` field:
+
+```yaml
```
Comment (pritidesai) [inline docs/pipelines.md:324]: will be great to have a similar example in `finally` and [custom tasks](https://github.com/tektoncd/pipeline/blob/e471b44fe38cb60b35ad8a064d95e0d97d9ce583/docs/pipelines.md#using-custom-tasks)

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go (hunk 1 above)
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipeline_validation.go:553]: can we instead implement this as a member function of `(pt PipelineTask)`?

**Comment by pritidesai** [comment]
thank you @jerop, its looking great, very excited 🎉  I have left a couple of comments 🙏

File: docs/pipelines.md (hunk 4 above)
Comment (jerop) [reply docs/pipelines.md:328]: great idea, added them :)

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go (hunk 1 above)
Comment (jerop) [reply pkg/apis/pipeline/v1beta1/pipeline_validation.go:553]: sure, refactored it :)

File: docs/matrix.md (hunk 2 above)
Comment (jerop) [reply docs/matrix.md:1]: 🤔 there are no copyright licenses on any other docs in https://github.com/tektoncd/pipeline/tree/main/docs (spot checked)

Comment (pritidesai) [reply docs/matrix.md:1]: Thanks @jerop for pointing that out, I have opened a github discussion for the same:

https://github.com/tektoncd/pipeline/discussions/4626

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go [hunk 5]
```diff
@@ -546,3 +549,19 @@ func validateGraph(tasks []PipelineTask) *apis.FieldError {
...
+	for idx, task := range tasks {
+		errs = errs.Also(task.validateMatrix(ctx).ViaIndex(idx))
+	}
+	return errs
+}
+
+func (pt *PipelineTask) validateMatrix(ctx context.Context) (errs *apis.FieldError) {
```
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipeline_validation.go:560]: NIT: the pipelineTask member functions are created in `pipeline_types.go` along with its definition:

https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/pipeline_types.go#L208

We can move this as a followup PR.

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go [hunk 6]
```diff
@@ -2571,6 +2572,72 @@ func TestPipelineTasksExecutionStatus(t *testing.T) {
 	}
 }
 
+// TestIncompatibleAPIVersions exercises validation of fields that
+// require a specific feature gate version in order to work.
+func TestIncompatibleAPIVersions(t *testing.T) {
```
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:2577]: NIT: this is testing just the matrix, suggest being explicit in the name 🙏

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go [hunk 7]
```diff
@@ -2571,6 +2572,72 @@ func TestPipelineTasksExecutionStatus(t *testing.T) {
...
+					"enable-api-fields": version,
+				})
+				cfg := &config.Config{
+					FeatureFlags: featureFlags,
+				}
+
+				ctx := config.ToContext(context.Background(), cfg)
+
+				ps.SetDefaults(ctx)
+				err := ps.Validate(ctx)
```
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:2627]: NIT: suggest calling `validateMatrix` here instead of `validate` since the intention of this unit test is to validate `matrix` specific functionality.

**Review by pritidesai — APPROVED** [review]
thank you @jerop, very exciting 🎉

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go (hunk 7 above)
Comment (jerop) [reply pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:2627]: would prefer to have it in the pipelinespec validation, so i can check against both `tasks` and `finally`, and to confirm that the `validateMatrix` is used in the pipeline validation

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go (hunk 6 above)
Comment (jerop) [reply pkg/apis/pipeline/v1beta1/pipeline_validation_test.go]: was thinking that we can support checking other alpha features in pipelines here, but makes sense to limit it to matrix for now until we have more alpha features in pipeline spec

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go (hunk 5 above)
Comment (jerop) [reply pkg/apis/pipeline/v1beta1/pipeline_validation.go]: nice! moved it, thanks :)

**Comment by sbwsg** [comment]
Cheers @jerop !

/lgtm
